x/emissions/keeper/msgserver/msg_server_demand.go

```diff
@@ -2,13 +2,21 @@ package msgserver
 
 import (
 	"context"
+	"time"
+
+	"github.com/allora-network/allora-chain/x/emissions/metrics"
 
 	"github.com/allora-network/allora-chain/x/emissions/types"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
-func (ms msgServer) FundTopic(ctx context.Context, msg *types.FundTopicRequest) (*types.FundTopicResponse, error) {
+func (ms msgServer) FundTopic(ctx context.Context, msg *types.FundTopicRequest,
+) (
+	_ *types.FundTopicResponse,
+	returnErr error,
```

Comment:
These errors will always be `nil` in your `defer` calls.  So you'll need to rework the whole PR unfortunately.

The reason: `defer` function arguments (and receivers, fwiw) are evaluated at the point where the `defer` call is first encountered in the code, and *not* when the deferred function actually runs (i.e., when leaving the outer function due to `panic` or `return`).

Check it out: https://go.dev/play/p/MQOXA_m1Z61 -- it's `nil`.  Same pattern, and the error is shown as `nil` by the `defer` function, but non-`nil` by the parent function.  This is because on the line where the `defer` happens, `err` is definitely `nil`.  It doesn't matter what `err` is when the function returns/panics.

However, you can still do what you're trying to do.  Here are two options:
1) by using a reference/pointer: https://go.dev/play/p/kQFWMC5BqTu
2) by using a closure: https://go.dev/play/p/EguVnE0g0CM

I propose it as a useful exercise for you to try to develop a way of expressing this logic that is correct, and also feels nice to use.  You'll learn a bit about the language in doing so.

Explaining why this behavior of the language is such a shitshow is probably out of scope for this review, but maybe we can have a quick internal call with everybody to talk about it.  This is not a fun little corner of Go!  It's complete trash!  So it's best to rely on patterns and an internal `utils` library for this, and maybe even a custom linter.  Very easy to get it wrong.  Even the most senior devs get it wrong all the time.

Reply:
I'll give you a hint, btw.  Here's a util func I carry around to every project I work on:
```go
func AddStack(err *error) {
	if *err != nil {
		*err = errors.WithStack(*err)
	}
}
```
To use it:
```go
func blah() (err error) {
    defer utils.AddStack(&err)
    
    err = doThing()
    if err != nil {
        return err
    }
}
```